File: AspNetMvcUnleashed/0672329980_source/Chapter14/VB/MvcApplication1/MvcApplication1/Controllers/SelectorController.vb
Public Class SelectorController
    Inherits Controller

    Private _entities As New ProductsDBEntities()

    Public Function Index()
        Dim categories = _entities.CategorySet.ToList()
        Dim products = New List(Of Product)()
        Return View(New ProductsVDM(categories, products))
    End Function

    <AcceptAjax(), ActionName("Details")> _
    Public Function Details_Uplevel(ByVal id As Integer) As ActionResult
        Dim products = From p In _entities.ProductSet _
                       Where p.CategoryId = id _
                       Select p

        Return PartialView("Details", products)
    End Function

    <ActionName("Details")> _
    Public Function Details_Downlevel(ByVal id As Integer) As ActionResult
        Dim categories = _entities.CategorySet.ToList()
        Dim products = From p In _entities.ProductSet _
                       Where p.CategoryId = id _
                       Select p

        Return View("Index", New ProductsVDM(categories, products))
    End Function

End Class

File: AspNetMvcUnleashed/0672329980_source/Chapter14/CS/MvcApplication1/MvcApplication1/Controllers/SelectorController.cs
using System.Collections.Generic;
using System.Linq;
using System.Web.Mvc;
using MvcApplication1.Models;

namespace MvcApplication1.Controllers
{

    public class SelectorController : Controller
    {
        private ProductsDBEntities _entities = new ProductsDBEntities();

        public ActionResult Index()
        {
            var categories = _entities.CategorySet.ToList();
            var products = new List<Product>();
            return View(new ProductsVDM(categories, products));
        }

        [AcceptAjax]
        [ActionName("Details")]
        public ActionResult Details_Uplevel(int id)
        {
            var products = from p in _entities.ProductSet
                           where p.CategoryId == id
                           select p;

            return PartialView("Details", products);
        }

        [ActionName("Details")]
        public ActionResult Details_Downlevel(int id)
        {
            var categories = _entities.CategorySet.ToList();         
            var products = from p in _entities.ProductSet
                           where p.CategoryId == id
                           select p;

            return View("Index", new ProductsVDM(categories, products));
        }



    }
}
